### PR TITLE
Closes #1366: Write test for FocusedDOMElementCache!

### LINKED ISSUE #1366: [1364] #2: Write test for FocusedDOMElementCache
This class fixes an issue where removing focus from the WebView and returning focus to the WebView resets the focused DOM element, which breaks navigation on some sites like YouTube TV. Focus is lost when opening the overlay. As such, these sites may be completely broken if this regresses. It's an edge case that isn't tested much and may not explicitly be covered by TestRail.

The test should roughly:
- Focus an element in the DOM (i.e. not the body element, which will gain focus when focus returns)
- Remove focus from the WebView
- Return focus to the WebView
- Verify the same element in the DOM is focused

### PR DESCRIPTION
**Pull request by mcomella**
Ignore commits from 1441 please.

### Pull Request checklist

- [x] This PR includes thorough **tests** or an explanation of why it does not
- [x] This PR includes a **CHANGELOG entry** or does not need one

### CODE REVIEW

File: app/src/androidTest/java/org/mozilla/tv/firefox/helpers/MainActivityTestRule.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import android.support.test.rule.ActivityTestRule
+import org.mozilla.tv.firefox.MainActivity
+import org.mozilla.tv.firefox.onboarding.OnboardingActivity.Companion.ONBOARD_SHOWN_PREF
+import org.mozilla.tv.firefox.pocket.PocketOnboardingActivity.Companion.POCKET_ONBOARDING_SHOWN_PREF
+
+/**
+ * A [org.junit.Rule] to handle shared test set up for tests on [MainActivity].
+ *
+ * @param initialTouchMode true to enable "touch mode", false otherwise
+ * @param launchActivity true to automatically launch the activity, false otherwise
```
Comment (Baron-Severin) [inline app/src/androidTest/java/org/mozilla/tv/firefox/helpers/MainActivityTestRule.kt]: Nit: this `launchActivity` description is unclear to me, but the description in `ActivityTestRule` is good.  Suggest either directing readers to it or poaching it entirely.

File: app/src/androidTest/java/org/mozilla/tv/firefox/ui/BasicNavigationTest.kt [hunk 2]
```diff
@@ -30,26 +24,9 @@ private val BODY_TEXT = (1..3).map {
...
-        loadingIdlingResource = SessionLoadedIdlingResource()
-        IdlingRegistry.getInstance().register(loadingIdlingResource)
-    }
-
-    @After
-    fun tearDown() {
-        IdlingRegistry.getInstance().unregister(loadingIdlingResource)
-        activityTestRule.activity.finishAndRemoveTask()
-    }
+    @Rule @JvmField val activityTestRule = MainActivityTestRule()
```
Comment (Baron-Severin) [inline app/src/androidTest/java/org/mozilla/tv/firefox/ui/BasicNavigationTest.kt]: See https://github.com/mozilla-mobile/firefox-tv/pull/1449#discussion_r230443548

File: app/src/androidTest/java/org/mozilla/tv/firefox/ui/DOMElementFocusTest.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import org.junit.Rule
+import org.junit.Test
+import org.mozilla.tv.firefox.ext.toUri
+import org.mozilla.tv.firefox.helpers.AndroidAssetDispatcher
+import org.mozilla.tv.firefox.helpers.MainActivityTestRule
+import org.mozilla.tv.firefox.ui.robots.BrowserRobot
+import org.mozilla.tv.firefox.ui.robots.navigationOverlay
+
+/**
+ * Ensures that, if the EngineView loses and regains focus, the focused DOM element maintains focused.
```
Comment (Baron-Severin) [inline app/src/androidTest/java/org/mozilla/tv/firefox/ui/DOMElementFocusTest.kt]: Nit: maintains focused. -> maintains focus.

File: app/src/androidTest/java/org/mozilla/tv/firefox/ui/DOMElementFocusTest.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ * If the focused element changes, sites like YouTube TV may break. See
+ * [org.mozilla.tv.firefox.webrender.FocusedDOMElementCacheInterface] for additional details on why this is necessary.
```
Comment (Baron-Severin) [inline app/src/androidTest/java/org/mozilla/tv/firefox/ui/DOMElementFocusTest.kt]: Nit: since we're describing the problem here anyway, is there a good issue # that could be included?

Comment (Baron-Severin) [inline app/src/androidTest/java/org/mozilla/tv/firefox/ui/DOMElementFocusTest.kt]: Optional nit: a top level import for `org.mozilla.tv.firefox.webrender.FocusedDOMElementCacheInterface` would make this easier to read.

Comment (Baron-Severin) [inline app/src/androidTest/java/org/mozilla/tv/firefox/ui/DOMElementFocusTest.kt]: There aren't any changes to FocusedDOMElementCacheInterface, but I notice that its kdoc includes links to classes that are neither fully qualified nor imported.  Could you add those, so that future renames don't break its documentation?

File: app/src/androidTest/java/org/mozilla/tv/firefox/ui/robots/BrowserRobot.kt [hunk 5]
```diff
@@ -16,6 +16,16 @@ import org.hamcrest.CoreMatchers.equalTo
 
 class BrowserRobot {
 
+    /**
+     * Executes the given JS string.
+     * @return if the final expression is a return statement, returns the provided String, else null.
+     */
+    fun executeJS(js: String): String? {
```
Comment (Baron-Severin) [inline app/src/androidTest/java/org/mozilla/tv/firefox/ui/robots/BrowserRobot.kt]: This is a neat helper function.

File: app/src/androidTest/java/org/mozilla/tv/firefox/ui/DOMElementFocusTest.kt (hunk 4 above)
Comment (mcomella) [reply app/src/androidTest/java/org/mozilla/tv/firefox/ui/DOMElementFocusTest.kt]: FocusedDOMElementCache has specifics on the problem as well as our GitHub bug number and the chromium bug number: I don't think it's worth repeating it here.
